wire/msgtx.go

```diff
@@ -336,9 +336,10 @@ func (msg *MsgTx) TxHash() chainhash.Hash {
 	// Ignore the error returns since the only way the encode could fail
 	// is being out of memory or due to nil pointers, both of which would
 	// cause a run-time panic.
-	buf := bytes.NewBuffer(make([]byte, 0, msg.SerializeSizeStripped()))
-	_ = msg.SerializeNoWitness(buf)
-	return chainhash.DoubleHashH(buf.Bytes())
+	txHash := sha256.New()
```

Comment:
One other optimization that I've found to really speed things up a lot of caching the result after the first instance. During validation, we end up recomputing this hash a few times: merkle tree check, sighash calculation, etc.

Reply:
`btcutil.Tx` caches the hash right? Maybe try to use that more instead of the raw `MsgTx`. Though I've not looked yet and maybe that's not possible.

Reply:
> btcutil.Tx caches the hash right?

Yeah that caches, but in most instances we use `wire.MsgTx` instead, mainly for historical reasons (but also because we decode from the wire into that struct). I was doing some syncing tests the other day and did some profiling of a catch up sync well past the checkpoints: https://github.com/btcsuite/btcd/pull/1376#issuecomment-1672326490